File: examples/canonical-auth/ruby/lib/tenantcore_auth/domain/in_memory_store.rb
require "json"
require "pathname"
require "securerandom"
require "set"

require_relative "models"

module TenantcoreAuth
  module Domain
    class InMemoryStore
      CANONICAL_PERMISSION_NAMES = [
        "iam:user:read",
        "iam:user:create",
        "iam:user:update",
        "iam:user:delete",
        "iam:user:invite",
        "iam:group:read",
        "iam:group:create",
        "iam:group:update",
        "iam:group:delete",
        "iam:group:assign_permission",
        "iam:group:assign_user",
        "iam:tenant:read",
        "iam:tenant:create",
        "iam:tenant:update",
        "iam:permission:read",
        "billing:invoice:read",
        "billing:invoice:create",
        "billing:subscription:read",
        "billing:subscription:update",
        "reports:usage:read",
        "reports:usage:export",
        "reports:audit:read",
        "admin:tenant:create",
        "admin:tenant:suspend",
        "admin:audit:read"
      ].freeze

      def self.default_fixture_dir
        Pathname.new(__dir__).join("..", "..", "..", "..", "domain", "fixtures").expand_path
      end

      def self.load_from_fixtures(fixture_dir = default_fixture_dir)
        fixture_path = Pathname.new(fixture_dir)
        users = load_collection(fixture_path.join("users.json"), User)
        tenants = load_collection(fixture_path.join("tenants.json"), Tenant)
        groups = load_collection(fixture_path.join("groups.json"), Group)
        permissions = load_collection(fixture_path.join("permissions.json"), Permission)
        memberships = load_collection(fixture_path.join("memberships.json"), Membership)
        group_permissions = load_collection(fixture_path.join("group_permissions.json"), GroupPermission)
        user_groups = load_collection(fixture_path.join("user_groups.json"), UserGroup)

        new(
          users: users,
          tenants: tenants,
          groups: groups,
          permissions: permissions,
          memberships: memberships,
          group_permissions: group_permissions,
          user_groups: user_groups
        ).tap(&:seed_canonical_permissions!)
      end

      def self.load_collection(path, struct_class)
        JSON.parse(path.read, symbolize_names: true).map do |row|
          normalized = row.transform_values do |value|
            value.is_a?(String) && value.match?(/\A\d{4}-\d{2}-\d{2}T/) ? Time.iso8601(value) : value
          end
          struct_class.new(**normalized)
        end
      end

      attr_reader :users, :tenants, :groups, :permissions, :memberships, :group_permissions, :user_groups

      def initialize(users:, tenants:, groups:, permissions:, memberships:, group_permissions:, user_groups:)
        @users = users
        @tenants = tenants
        @groups = groups
        @permissions = permissions
        @memberships = memberships
        @group_permissions = group_permissions
        @user_groups = user_groups
        refresh_indexes!
      end

      def get_user_by_id(id)
        @users_by_id[id]
      end

      def get_user_by_email(email)
        @users_by_email[email]
      end

      def get_tenant_by_id(id)
        @tenants_by_id[id]
      end

      def get_tenant_name(id)
        get_tenant_by_id(id)&.name
      end

      def get_group_by_id(id)
        @groups_by_id[id]
      end

      def get_permission_by_name(name)
        @permissions_by_name[name]
      end

      def get_active_membership(user_id)
        @memberships.find { |membership| membership.user_id == user_id && membership.is_active }
      end

      def verify_membership(user_id, tenant_id)
        @memberships.any? do |membership|
          membership.user_id == user_id && membership.tenant_id == tenant_id && membership.is_active
        end
      end

      def get_membership_role(user_id, tenant_id = nil)
        membership = @memberships.find do |row|
          row.user_id == user_id && row.tenant_id == tenant_id && row.is_active
        end
        membership&.tenant_role
      end

      def get_groups_for_user(user_id, tenant_id)
        group_ids = @user_groups.select { |row| row.user_id == user_id }.map(&:group_id)
        group_ids.map do |group_id|
          group = @groups_by_id[group_id]
          group if group && group.tenant_id == tenant_id
        end.compact
      end

      def get_effective_permissions(user_id)
        membership = get_active_membership(user_id)
        return [] unless membership

        case membership.tenant_role
        when "tenant_admin"
          @permissions.map(&:name).sort
        when "super_admin"
          admin_permissions
        else
          group_ids = @user_groups.select { |row| row.user_id == user_id }.map(&:group_id).to_set
          @group_permissions.map do |row|
            @permissions_by_id[row.permission_id]&.name if group_ids.include?(row.group_id)
          end.compact.uniq.sort
        end
      end

      def admin_permissions
        @permissions.map(&:name).grep(/\Aadmin:/).sort
      end

      def list_users(tenant_id)
        @users.select { |user| user.tenant_id == tenant_id }.sort_by(&:email)
      end

      def list_groups(tenant_id)
        @groups.select { |group| group.tenant_id == tenant_id }.sort_by(&:slug)
      end

      def list_permissions
        @permissions.sort_by(&:name)
      end

      def list_tenants
        @tenants.sort_by(&:slug)
      end

      def invite_user(tenant_id:, email:, display_name:, group_slugs: [])
        user = User.new(
          id: "u-#{SecureRandom.uuid}",
          email: email,
          display_name: display_name,
          tenant_id: tenant_id,
          created_at: Time.now.utc,
          is_active: true,
          acl_ver: 1
        )
        @users << user
        @memberships << Membership.new(
          id: "m-#{SecureRandom.uuid}",
          user_id: user.id,
          tenant_id: tenant_id,
          tenant_role: "tenant_member",
          created_at: Time.now.utc,
          is_active: true
        )
        group_slugs.each do |slug|
          group = @groups.find { |row| row.slug == slug && row.tenant_id == tenant_id }
          next unless group

          @user_groups << UserGroup.new(
            id: "ug-#{SecureRandom.uuid}",
            user_id: user.id,
            group_id: group.id,
            joined_at: Time.now.utc
          )
        end
        refresh_indexes!
        user
      end

      def create_group(tenant_id:, slug:, name:, permission_names: [])
        group = Group.new(
          id: "g-#{SecureRandom.uuid}",
          tenant_id: tenant_id,
          slug: slug,
          name: name,
          created_at: Time.now.utc
        )
        @groups << group
        permission_names.each do |permission_name|
          assign_permission_to_group(group.id, permission_name, tenant_id)
        end
        refresh_indexes!
        group
      end

      def assign_permission_to_group(group_id, permission_name, tenant_id)
        group = @groups_by_id[group_id]
        permission = @permissions_by_name[permission_name]
        return false unless group && permission
        return false unless group.tenant_id == tenant_id

        already_assigned = @group_permissions.any? do |row|
          row.group_id == group_id && row.permission_id == permission.id
        end
        return true if already_assigned

        @group_permissions << GroupPermission.new(
          id: "gp-#{SecureRandom.uuid}",
          group_id: group_id,
          permission_id: permission.id,
          granted_at: Time.now.utc
        )
        true
      end

      def assign_user_to_group(group_id, user_id, tenant_id)
        group = @groups_by_id[group_id]
        user = @users_by_id[user_id]
        return false unless group && user
        return false unless group.tenant_id == tenant_id && user.tenant_id == tenant_id

        already_assigned = @user_groups.any? { |row| row.group_id == group_id && row.user_id == user_id }
        return true if already_assigned

        @user_groups << UserGroup.new(
          id: "ug-#{SecureRandom.uuid}",
          user_id: user_id,
          group_id: group_id,
          joined_at: Time.now.utc
        )
        true
      end

      def create_tenant(slug:, name:, first_admin_email:)
        tenant = Tenant.new(
          id: "t-#{SecureRandom.uuid}",
          slug: slug,
          name: name,
          created_at: Time.now.utc,
          is_active: true
        )
        @tenants << tenant
        admin = User.new(
          id: "u-#{SecureRandom.uuid}",
          email: first_admin_email,
          display_name: "#{name} Admin",
          tenant_id: tenant.id,
          created_at: Time.now.utc,
          is_active: true,
          acl_ver: 1
        )
        @users << admin
        @memberships << Membership.new(
          id: "m-#{SecureRandom.uuid}",
          user_id: admin.id,
          tenant_id: tenant.id,
          tenant_role: "tenant_admin",
          created_at: Time.now.utc,
          is_active: true
        )
        refresh_indexes!
        tenant
      end

      def seed_canonical_permissions!
        existing = @permissions.map(&:name).to_set
        CANONICAL_PERMISSION_NAMES.each do |permission_name|
          next if existing.include?(permission_name)

          @permissions << Permission.new(
            id: "perm-#{permission_name.tr(':', '-')}",
            name: permission_name,
            description: "Canonical seeded permission",
            created_at: Time.iso8601("2025-01-01T00:00:00Z")
          )
        end
        refresh_indexes!
      end

      private

      def refresh_indexes!
        @users_by_id = @users.to_h { |row| [row.id, row] }
        @users_by_email = @users.to_h { |row| [row.email, row] }
        @tenants_by_id = @tenants.to_h { |row| [row.id, row] }
        @groups_by_id = @groups.to_h { |row| [row.id, row] }
        @permissions_by_id = @permissions.to_h { |row| [row.id, row] }
        @permissions_by_name = @permissions.to_h { |row| [row.name, row] }
      end
    end
  end
end

File: examples/canonical-auth/elixir/lib/tenantcore_auth/domain/in_memory_store.ex
defmodule TenantcoreAuth.Domain.InMemoryStore do
  use GenServer

  alias TenantcoreAuth.Domain.{Group, GroupPermission, Membership, Permission, Tenant, User, UserGroup}

  @canonical_permissions [
    "iam:user:read",
    "iam:user:create",
    "iam:user:update",
    "iam:user:delete",
    "iam:user:invite",
    "iam:group:read",
    "iam:group:create",
    "iam:group:update",
    "iam:group:delete",
    "iam:group:assign_permission",
    "iam:group:assign_user",
    "iam:tenant:read",
    "iam:tenant:create",
    "iam:tenant:update",
    "iam:permission:read",
    "billing:invoice:read",
    "billing:invoice:create",
    "billing:subscription:read",
    "billing:subscription:update",
    "reports:usage:read",
    "reports:usage:export",
    "reports:audit:read",
    "admin:tenant:create",
    "admin:tenant:suspend",
    "admin:audit:read"
  ]

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @impl true
  def init(_opts) do
    {:ok, load_fixtures(default_fixture_dir())}
  end

  def get_user_by_id(id), do: GenServer.call(__MODULE__, {:get_user_by_id, id})
  def get_user_by_email(email), do: GenServer.call(__MODULE__, {:get_user_by_email, email})
  def get_tenant_name(id), do: GenServer.call(__MODULE__, {:get_tenant_name, id})
  def get_groups_for_user(user_id, tenant_id), do: GenServer.call(__MODULE__, {:get_groups_for_user, user_id, tenant_id})
  def get_effective_permissions(user_id), do: GenServer.call(__MODULE__, {:get_effective_permissions, user_id})
  def verify_membership(user_id, tenant_id), do: GenServer.call(__MODULE__, {:verify_membership, user_id, tenant_id})
  def get_active_membership(user_id), do: GenServer.call(__MODULE__, {:get_active_membership, user_id})
  def list_users(tenant_id), do: GenServer.call(__MODULE__, {:list_users, tenant_id})
  def list_groups(tenant_id), do: GenServer.call(__MODULE__, {:list_groups, tenant_id})
  def list_permissions, do: GenServer.call(__MODULE__, :list_permissions)
  def list_tenants, do: GenServer.call(__MODULE__, :list_tenants)
  def get_group_by_id(id), do: GenServer.call(__MODULE__, {:get_group_by_id, id})

  def invite_user(tenant_id, email, display_name, group_slugs \\ []) do
    GenServer.call(__MODULE__, {:invite_user, tenant_id, email, display_name, group_slugs})
  end

  def create_group(tenant_id, slug, name, permission_names \\ []) do
    GenServer.call(__MODULE__, {:create_group, tenant_id, slug, name, permission_names})
  end

  def assign_permission_to_group(group_id, permission_name, tenant_id) do
    GenServer.call(__MODULE__, {:assign_permission_to_group, group_id, permission_name, tenant_id})
  end

  def assign_user_to_group(group_id, user_id, tenant_id) do
    GenServer.call(__MODULE__, {:assign_user_to_group, group_id, user_id, tenant_id})
  end

  def create_tenant(slug, name, first_admin_email) do
    GenServer.call(__MODULE__, {:create_tenant, slug, name, first_admin_email})
  end

  @impl true
  def handle_call({:get_user_by_id, id}, _from, state), do: {:reply, Map.get(state.users_by_id, id), state}
  def handle_call({:get_user_by_email, email}, _from, state), do: {:reply, Map.get(state.users_by_email, email), state}
  def handle_call({:get_tenant_name, id}, _from, state), do: {:reply, get_in(state.tenants_by_id, [id, Access.key(:name)]), state}
  def handle_call({:get_group_by_id, id}, _from, state), do: {:reply, Map.get(state.groups_by_id, id), state}

  def handle_call({:get_groups_for_user, user_id, tenant_id}, _from, state) do
    group_ids = state.user_groups |> Enum.filter(&(&1.user_id == user_id)) |> Enum.map(& &1.group_id) |> MapSet.new()
    groups =
      state.groups
      |> Enum.filter(&MapSet.member?(group_ids, &1.id))
      |> Enum.filter(&(&1.tenant_id == tenant_id))
    {:reply, groups, state}
  end

  def handle_call({:get_effective_permissions, user_id}, _from, state) do
    membership = Enum.find(state.memberships, &(&1.user_id == user_id and &1.is_active))

    permissions =
      case membership && membership.tenant_role do
        "tenant_admin" ->
          state.permissions |> Enum.map(& &1.name) |> Enum.sort()

        "super_admin" ->
          state.permissions |> Enum.map(& &1.name) |> Enum.filter(&String.starts_with?(&1, "admin:")) |> Enum.sort()

        _ ->
          group_ids = state.user_groups |> Enum.filter(&(&1.user_id == user_id)) |> Enum.map(& &1.group_id) |> MapSet.new()

          state.group_permissions
          |> Enum.filter(&MapSet.member?(group_ids, &1.group_id))
          |> Enum.map(&Map.get(state.permissions_by_id, &1.permission_id))
          |> Enum.reject(&is_nil/1)
          |> Enum.map(& &1.name)
          |> Enum.uniq()
          |> Enum.sort()
      end

    {:reply, permissions, state}
  end

  def handle_call({:verify_membership, user_id, tenant_id}, _from, state) do
    result = Enum.any?(state.memberships, &(&1.user_id == user_id and &1.tenant_id == tenant_id and &1.is_active))
    {:reply, result, state}
  end

  def handle_call({:get_active_membership, user_id}, _from, state) do
    {:reply, Enum.find(state.memberships, &(&1.user_id == user_id and &1.is_active)), state}
  end

  def handle_call({:list_users, tenant_id}, _from, state) do
    {:reply, state.users |> Enum.filter(&(&1.tenant_id == tenant_id)) |> Enum.sort_by(& &1.email), state}
  end

  def handle_call({:list_groups, tenant_id}, _from, state) do
    {:reply, state.groups |> Enum.filter(&(&1.tenant_id == tenant_id)) |> Enum.sort_by(& &1.slug), state}
  end

  def handle_call(:list_permissions, _from, state), do: {:reply, Enum.sort_by(state.permissions, & &1.name), state}
  def handle_call(:list_tenants, _from, state), do: {:reply, Enum.sort_by(state.tenants, & &1.slug), state}

  def handle_call({:invite_user, tenant_id, email, display_name, group_slugs}, _from, state) do
    user = %User{id: "u-#{unique_id()}", email: email, display_name: display_name, tenant_id: tenant_id, created_at: now_iso(), is_active: true, acl_ver: 1}
    membership = %Membership{id: "m-#{unique_id()}", user_id: user.id, tenant_id: tenant_id, tenant_role: "tenant_member", created_at: now_iso(), is_active: true}

    user_groups =
      group_slugs
      |> Enum.map(fn slug -> Enum.find(state.groups, &(&1.slug == slug and &1.tenant_id == tenant_id)) end)
      |> Enum.reject(&is_nil/1)
      |> Enum.map(fn group -> %UserGroup{id: "ug-#{unique_id()}", user_id: user.id, group_id: group.id, joined_at: now_iso()} end)

    new_state = rebuild_state(%{
      state
      | users: [user | state.users],
        memberships: [membership | state.memberships],
        user_groups: user_groups ++ state.user_groups
    })

    {:reply, user, new_state}
  end

  def handle_call({:create_group, tenant_id, slug, name, permission_names}, _from, state) do
    group = %Group{id: "g-#{unique_id()}", tenant_id: tenant_id, slug: slug, name: name, created_at: now_iso()}

    group_permissions =
      permission_names
      |> Enum.map(&Map.get(state.permissions_by_name, &1))
      |> Enum.reject(&is_nil/1)
      |> Enum.map(fn permission ->
        %GroupPermission{id: "gp-#{unique_id()}", group_id: group.id, permission_id: permission.id, granted_at: now_iso()}
      end)

    new_state = rebuild_state(%{state | groups: [group | state.groups], group_permissions: group_permissions ++ state.group_permissions})
    {:reply, group, new_state}
  end

  def handle_call({:assign_permission_to_group, group_id, permission_name, tenant_id}, _from, state) do
    group = Map.get(state.groups_by_id, group_id)
    permission = Map.get(state.permissions_by_name, permission_name)

    cond do
      is_nil(group) or is_nil(permission) or group.tenant_id != tenant_id ->
        {:reply, false, state}

      Enum.any?(state.group_permissions, &(&1.group_id == group_id and &1.permission_id == permission.id)) ->
        {:reply, true, state}

      true ->
        new_row = %GroupPermission{id: "gp-#{unique_id()}", group_id: group_id, permission_id: permission.id, granted_at: now_iso()}
        {:reply, true, rebuild_state(%{state | group_permissions: [new_row | state.group_permissions]})}
    end
  end

  def handle_call({:assign_user_to_group, group_id, user_id, tenant_id}, _from, state) do
    group = Map.get(state.groups_by_id, group_id)
    user = Map.get(state.users_by_id, user_id)

    cond do
      is_nil(group) or is_nil(user) or group.tenant_id != tenant_id or user.tenant_id != tenant_id ->
        {:reply, false, state}

      Enum.any?(state.user_groups, &(&1.group_id == group_id and &1.user_id == user_id)) ->
        {:reply, true, state}

      true ->
        row = %UserGroup{id: "ug-#{unique_id()}", user_id: user_id, group_id: group_id, joined_at: now_iso()}
        {:reply, true, rebuild_state(%{state | user_groups: [row | state.user_groups]})}
    end
  end

  def handle_call({:create_tenant, slug, name, first_admin_email}, _from, state) do
    tenant = %Tenant{id: "t-#{unique_id()}", slug: slug, name: name, created_at: now_iso(), is_active: true}
    admin = %User{id: "u-#{unique_id()}", email: first_admin_email, display_name: "#{name} Admin", tenant_id: tenant.id, created_at: now_iso(), is_active: true, acl_ver: 1}
    membership = %Membership{id: "m-#{unique_id()}", user_id: admin.id, tenant_id: tenant.id, tenant_role: "tenant_admin", created_at: now_iso(), is_active: true}

    new_state =
      rebuild_state(%{
        state
        | tenants: [tenant | state.tenants],
          users: [admin | state.users],
          memberships: [membership | state.memberships]
      })

    {:reply, tenant, new_state}
  end

  def default_fixture_dir do
    Path.expand("../../../../domain/fixtures", __DIR__)
  end

  defp load_fixtures(dir) do
    users = load_json(dir, "users.json", User)
    tenants = load_json(dir, "tenants.json", Tenant)
    groups = load_json(dir, "groups.json", Group)
    permissions = load_json(dir, "permissions.json", Permission)
    memberships = load_json(dir, "memberships.json", Membership)
    group_permissions = load_json(dir, "group_permissions.json", GroupPermission)
    user_groups = load_json(dir, "user_groups.json", UserGroup)

    permissions =
      Enum.reduce(@canonical_permissions, permissions, fn name, acc ->
        if Enum.any?(acc, &(&1.name == name)) do
          acc
        else
          [%Permission{id: "perm-#{String.replace(name, ":", "-")}", name: name, description: "Canonical seeded permission", created_at: "2025-01-01T00:00:00Z"} | acc]
        end
      end)

    rebuild_state(%{
      users: users,
      tenants: tenants,
      groups: groups,
      permissions: permissions,
      memberships: memberships,
      group_permissions: group_permissions,
      user_groups: user_groups
    })
  end

  defp load_json(dir, filename, struct_module) do
    dir
    |> Path.join(filename)
    |> File.read!()
    |> Jason.decode!()
    |> Enum.map(&struct(struct_module, atomize_keys(&1)))
  end

  defp atomize_keys(map) do
    Map.new(map, fn {key, value} -> {String.to_atom(key), value} end)
  end

  defp rebuild_state(state) do
    Map.merge(state, %{
      users_by_id: Map.new(state.users, &{&1.id, &1}),
      users_by_email: Map.new(state.users, &{&1.email, &1}),
      tenants_by_id: Map.new(state.tenants, &{&1.id, &1}),
      groups_by_id: Map.new(state.groups, &{&1.id, &1}),
      permissions_by_id: Map.new(state.permissions, &{&1.id, &1}),
      permissions_by_name: Map.new(state.permissions, &{&1.name, &1})
    })
  end

  defp unique_id, do: System.unique_integer([:positive]) |> Integer.to_string()
  defp now_iso, do: DateTime.utc_now() |> DateTime.truncate(:second) |> DateTime.to_iso8601()
end
